### ISSUE TITLE
[Good First Issue]: Align behavior of ONNX Frontend function ReduceSumSquare-11, 13, 18 with original framework

### ISSUE THREAD

**Issue by gkrivor** [openvinotoolkit/openvino#20563]
### Context

Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.
OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.
This task requires alignment between OpenVINO ONNX Frontend and original framework implementations of ReduceSumSquare for next list of opsets: [opset 11](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-11),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-13),  [opset 18](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-18)
Necessary help will be provided by ONNX Fronted team.

### What needs to be done?

First of all, please, take a look on [ReduceMax](https://github.com/openvinotoolkit/openvino/pull/23475) PR for a reference.

Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md#ReduceSumSquare)
More details can be found in ONNX Changelog: [opset 11](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-11),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-13),  [opset 18](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceSumSquare-18)
1. Function already has a common implementation in OpenVINO. First of all, you need to review a documentation and prepare a table with differences between versions. It could be, for instance, a missing property, extended/reduced coverage of existing property, etc...
2. Copy existing implementation [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op) to make it aligned with original framework (extend or reduce coverage of a common implementation). Copy of modified implementation should be in a defined opset, or in opset 1 in case it implements oldest implementation. [Example of multi-opset operation](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp#L50).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) to find a test marked as a xfailed for added functionality. If any exist - remove corresponding lines and try to verify by using cmdline "python -m pytest -k name_of_test".
More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/wPuqAujS) - engage in discussions, ask questions and talk to OpenVINO developers
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [User documentation](https://docs.openvino.ai/)
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

### Contact points

@gkrivor

**Comment by inbasperu** [comment]
.take

**Comment by inbasperu** [comment]
Hello OpenVINO maintainers,

I've set up and built the OpenVINO repository locally and have been exploring potential solutions for the issue at hand. Below is my proposed approach, and I would greatly appreciate any feedback or guidance on whether I'm heading in the right direction.

### Proposed Approach

1. **Differences Table**: I've reviewed the changelogs and compiled a table outlining the differences between the three specified versions.

| Feature/Property                   | Opset 11                                                                                   | Opset 13                                                                              | Opset 18                                                                                                      |
| ---------------------------------- | ------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------- |
| **Attributes**                     | - axes: list of ints<br>- keepdims: int (default 1)                                        | - axes: list of ints<br>- keepdims: int (default 1)                                   | - keepdims: int (default 1)<br>- noop_with_empty_axes: int (default 0)                                        |
| **Axes Attribute**                 | A list of integers to reduce along range:`[-r, r-1] `. Default is all dimensions.          | A list of integers to reduce along range:`[-r, r-1] `. Default is all dimensions.     | Not present                                                                                                   |
| **Keepdims Attribute**             | Keep the reduced dimension (default is to keep).                                           | Keep the reduced dimension (default is to keep).                                      | Keep the reduced dimension (default is to keep).                                                              |
| **Noop_with_empty_axes Attribute** | Not present                                                                                | Not present                                                                           | Defines behavior if 'axes' is empty. Reduces all axes by default if false, or acts as an identity op if true. |
| **Inputs**                         | - data: T                                                                                  | - data (differentiable): T                                                            | - data (differentiable): T<br>- axes (optional, non-differentiable): tensor(int64)                            |
| **Input Tensors of Rank Zero**     | Not specified                                                                              | Valid input tensors                                                                   | Valid input tensors                                                                                           |
| **Reduction over Empty Set**       | Not specified                                                                              | Yields 0                                                                              | Yields 0                                                                                                      |
| **Outputs**                        | - reduced: T                                                                               | - reduced (differentiable): T                                                         | - reduced (differentiable): T                                                                                 |
| **Type Constraints**               | - T: high-precision numeric tensors (uint32, uint64, int32, int64, float16, float, double) | - T: numeric tensors (uint32, uint64, int32, int64, float16, float, double, bfloat16) | - T: numeric tensors (uint32, uint64, int32, int64, float16, float, double, bfloat16)                         |

Key differences between versions:

- The `noop_with_empty_axes` attribute is introduced in Opset 18, which specifies the behavior when the `axes` attribute is empty.
- Opset 13 started to mention that reduction over an empty set of values yields 0 and that input tensors of rank zero are valid, which was not specified in Opset 11.
- The type constraint was extended in Opset 13 to include `tensor(bfloat16)`, which was not supported in earlier versions.
- In Opset 18, the `axes` attribute becomes an optional input, whereas in Opsets 11 and 13, it is a required attribute with a default value.

2. **Common Implementation for Multi Opset**:
   - **Files to Modify**:
     - Header: [reduce.hpp](https://github.com/openvinotoolkit/openvino/blob/05f498db303ed13a1224a44d71d124a178208f44/src/frontends/onnx/frontend/src/op/reduce.hpp#L167)
     - Implementation: [reduce.cpp](https://github.com/openvinotoolkit/openvino/blob/6accdd741c2a4141415714381ef70121c25dd693/src/frontends/onnx/frontend/src/op/reduce.cpp#L154)
   - Goal: Align behavior across the specified versions.

3. **Function Registration**:
   - Location: [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/6accdd741c2a4141415714381ef70121c25dd693/src/frontends/onnx/frontend/src/ops_bridge.cpp#L488)

4. **Test Models**:
   - Add prototext models in the [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/), referencing an existing model like [reduce_sum_square.prototxt](https://github.com/openvinotoolkit/openvino/blob/6accdd741c2a4141415714381ef70121c25dd693/src/frontends/onnx/tests/models/reduce_sum_square.prototxt#L4).

5. **Use Case Coverage**:
   - Update [onnx_import.in.cpp](https://github.com/openvinotoolkit/openvino/blob/6accdd741c2a4141415714381ef70121c25dd693/src/frontends/onnx/tests/onnx_import.in.cpp#L1057) with tests for all use cases of the ONNX models.

6. **Python xfailed Tests**:
   - Review [test_backend.py](https://github.com/openvinotoolkit/openvino/blob/6accdd741c2a4141415714381ef70121c25dd693/src/frontends/onnx/tests/tests_python/test_backend.py#L526) for any tests marked as xfailed for the added functionality, remove the corresponding lines, and validate using the command `python -m pytest -k name_of_test`.

### Additional Context

I've also identified similar issues that could offer insights into this problem:
- [#20554](https://github.com/openvinotoolkit/openvino/issues/20554)
- [#20559](https://github.com/openvinotoolkit/openvino/issues/20559)
- [#20556](https://github.com/openvinotoolkit/openvino/issues/20556)
- [#20553](https://github.com/openvinotoolkit/openvino/issues/20553)
- [#22075](https://github.com/openvinotoolkit/openvino/pull/22075)

For further understanding of multi-opset operations, I studied the softmax function implementation and have compiled my notes [here](https://docs.google.com/document/d/116J7mbWT6bm6kF9WAsEEI8qVeFZiZiIOO_5P67QiZRw/edit?usp=sharing).

I'm eager to contribute to the resolution of this issue and look forward to your feedback to ensure my efforts are aligned with the project's needs.

**Comment by p-wysocki** [comment]
Hello @inbasperu, the work you've done so far is very impressive, looks good to me. I think we could start with the implementation. :)

**Comment by inbasperu** [comment]
Hello, 
During testing in my development environment, the test suite failed to execute successfully, despite no recent changes being made to the repository. The tests are attempting to access specific ONNX model files, such as `quantize_linear_const.onnx`, `quantize_linear_u16.onnx`, and `quantize_linear_zero_point.onnx`, but cannot locate them at the specified paths.

The following error message was encountered:
```
unknown file: Failure
C++ exception with description "Exception from src/frontends/onnx/onnx_common/src/parser.cpp:24:
Could not open the file: /Users/inbasekaranperumal/Developer/OpenSource/openvino/bin/arm64/Release/./test_model_zoo/onnx/quantize_linear_const.onnx"
```
![image](https://github.com/openvinotoolkit/openvino/assets/158610103/80820a3a-af73-4a74-85d8-1e3eb4ece4ab)


Are these ONNX files supposed to be generated during the build or setup process?

## Build Instructions

```bash
cmake -G "Ninja Multi-Config" \
      -DCMAKE_BUILD_TYPE=Debug \
      -DENABLE_PYTHON=ON \
      -DENABLE_TESTS=ON \
      -DENABLE_FUNCTIONAL_TESTS=ON \
      -DENABLE_DEBUG_CAPS=ON \
      -DENABLE_CPU_DEBUG_CAPS=ON \
      -DENABLE_NCC_STYLE=ON \
      -DENABLE_SYSTEM_PUGIXML=ON \
      -DENABLE_SYSTEM_SNAPPY=ON \
      -DENABLE_SYSTEM_PROTOBUF=ON \
      -DPython3_EXECUTABLE=.venv/bin/python3 ..

cmake --build . --parallel $(sysctl -n hw.ncpu)
```

Please provide guidance on resolving this issue.

## System Information

- **Device:** M3 Pro
- **Operating System:** macOS Sonoma 14.2.1

**Comment by inbasperu** [comment]
## Linking Error with `libtbb_debug.12.2.dylib` on macOS in Debug Configuration

While attempting to build a project in Debug configuration on macOS, I encountered a linker error concerning `libtbb_debug.12.2.dylib`. The error suggests a problem with the `__LINKEDIT` segment of the TBB library file, preventing successful compilation.

I configure the project with CMake for a Debug build with the following execute the build command: `cmake --build . --config Debug --parallel $(sysctl -n hw.ncpu)`

### Error Message
The build process fails with the following linker error:
<img width="1435" alt="image" src="https://github.com/openvinotoolkit/openvino/assets/158610103/eb5419ea-9201-4d2c-9f5d-5ca81e113e92">
<img width="1441" alt="image" src="https://github.com/openvinotoolkit/openvino/assets/158610103/841c119d-e032-43f1-a5fa-b962fc2d6a1e">

```
ld: segment '__LINKEDIT' load command content extends beyond end of file in '/Users/inbasekaranperumal/Developer/OpenSource/openvino/temp/tbb/lib/libtbb_debug.12.2.dylib'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Environment
- **Device:** M3 Pro
- **Operating System:** macOS Sonoma 14.2.1
- **Compiler:** Apple clang version 15.0.0
- **TBB Version:** 12.2 (debug version)

- The issue specifically occurs with the debug version of the TBB library.
- I have confirmed the existence of the `libtbb_debug.12.2.dylib` file and attempted to rebuild it, yet the problem persists.

I'm seeking advice to resolve this linking issue. Any input would be greatly appreciated. Thanks in advance for your time and help.

**Comment by inbasperu** [comment]
Hello @p-wysocki,

I have started the implementation and encountered some uncertainties along the way. I have made a [PR](https://github.com/openvinotoolkit/openvino/pull/22993#issue-2146681772) with the changes I've added. It would be great if you could review it and guide me on how to proceed further.

**Comment by p-wysocki** [comment]
Thank you for the PR! Let's continue the discussion there.
